### ISSUE TITLE
A question beyond the code itself

### ISSUE THREAD

**Issue by RobinWenqian** [sgrvinod/a-PyTorch-Tutorial-to-Image-Captioning#11]
Under the help of you few days ago, I successfully run the code and get the results. Thank you again!

This time I wanna ask several questions beyond the code itself:
1. During the training process, is the supervision loss only cross-entropy loss?
2. During the training process, for each input image(carrying label), do we have the generated caption by the network? 
3. After the question 2, if I have the generated caption, how can I search for a similar caption and its corresponding image(or several) from the dataset? From your experience, what is the quickest way to accomplish this task?

Sincerely hope for your reply! If I there is anything unclear, we might discuss under this issue.
Thank you again for your help!

**Comment by sgrvinod** [comment]
Hi @RobinWenqian, sure,

1. No, the supervision loss is the sum of the cross-entropy loss and the ['doubly stochastic' regularization loss](https://github.com/sgrvinod/a-PyTorch-Tutorial-to-Image-Captioning/blob/b0191b381ab5ee6970921c771dba57f16bc09a46/train.py#L186), as described [here](https://github.com/sgrvinod/a-PyTorch-Tutorial-to-Image-Captioning#loss-function). This regularization encourages the model to attend to every part of the image over the course of the caption generation -- we punish it if it doesn't. In the paper, the authors state,

> In our experiments, we observed that this penalty was important quantitatively to improving overall BLEU score and that qualitatively this leads to more rich and descriptive captions

2. Because we are **Teacher Forcing** during training, i.e. providing the ground truth at each decode-step _regardless_ of the previously generated word, we are effectively only generating single 'next' words for each ground truth 'current' word. So we don't strictly have chain-generated captions, but, yes, you could assume that these 'next' words strung together form captions.

3. Evaluating similarity is often not a straightforward task in NLP, and takes some experimenting to find out what works best for the definition of similarity you have in mind. To start, you could calculate the BLEU score of the generated caption with respect to each set of `N_c` ground truth captions of each image. You could then consider all images/caption-sets that have a BLEU score above a threshold to be 'similar'. This would be easiest, since we're [already using](https://github.com/sgrvinod/a-PyTorch-Tutorial-to-Image-Captioning/blob/b0191b381ab5ee6970921c771dba57f16bc09a46/train.py#L316) the BLEU score. If you want to consider the _meanings_ of the words, you would need to use word embeddings -- for example, to encode each caption as the sum of the embeddings of the words, and then use a similarity metric like the cosine distance. Alternatively, there are other less naive but easy-to-use methods to encode your captions [like this one](https://github.com/facebookresearch/InferSent#use-our-sentence-encoder) that are proven to [work well for benchmark semantic similarity (STS) tasks](https://github.com/facebookresearch/InferSent#evaluate-the-encoder-on-transfer-tasks).

**Comment by RobinWenqian** [comment]
Really thank you for your patient reply! There is a follow-up question after question3.
You say **'You could then consider all images/caption-sets that have a BLEU score above a threshold to be 'similar''** Does it mean in the training process, we need to calculate BLEU score **for each generated caption with all the captions in the dataset**?

**Comment by sgrvinod** [comment]
For each generated caption with all **caption-sets** in the dataset. A single set contains `N_c` captions, and each image has one set. That is, if you have 100 images in the dataset with 5 (`N_c`) captions each, then you will calculate the BLEU score for the generated caption with respect to each of the 100 sets of 5 captions each. 

For each image/caption-set,

`sentence_bleu(references=[gt1, gt2, gt3, gt4, gt5], hypothesis=generated_caption)`

or

`corpus_bleu(references=[[gt1, gt2, gt3, gt4, gt5]], hypothesis=[generated_caption])`

The above lines are equivalent, see [here](https://www.nltk.org/_modules/nltk/translate/bleu_score.html) for more details on usage.

You could also try doing it for each caption instead of each caption-set. But as I mentioned earlier, the meanings of words are not considered in the BLEU score, so two words like `aeroplane` and `plane` are considered to be different words even though they mean the same. Using multiple (`N_c`) ground truth captions in each set somewhat mitigates this, since there is a chance that both words might be used across the `N_c` captions.

I'm curious - can I ask why you would like to find similar captions/images during the training process? Especially because captions are probably not well-formed during training. Once they are well formed, they might essentially be the same as the ground truth label, so it might be the same if you found similar captions for each ground truth caption in the dataset.

**Comment by RobinWenqian** [comment]
Oh no problem. Actually I need to find similar caption for each ground-truth caption during training process because I am working at **generating discriminating captions for similar images**. So my thought is to search for captions similar to ground-truth captions (in the dataset similar captions indicate images with them are similar) in the dataset. Then **push generated captions away from similar captions for a certain 'distance' while shorten the distance between generated captions and ground-truth** captions. The work may be concluded as adding a process (search for similar captions and push away) to most image-captioning framework during training in order to generate more discriminating captions.

From your experience, is this thought meaningful and if it is, should the search-push process start at very first of training process or after certain iterations of training?

Sincerely appreciate for your reply.

**Comment by anyface** [comment]
Very good！

**Comment by sgrvinod** [comment]
@RobinWenqian I see, that's interesting. It seems meaningful to me, although it's hard to say how well it would work. 

If I understand you right, you want the model to be more discriminating when it comes to images that are about similar subjects but yet have notable differences. Because image captioning models often generate a more 'general' caption which is broadly correct but _without_ more specific details as a human would provide. 

I haven't given it much thought, but one thing you could try with your approach is to find words that usually discriminate between similar images. (These may be words that point to specific details, like color, etc.) You could penalize the model more than usual when it misses out on these _discriminating_ words by assigning them a higher weight in the loss function, using the [`weight`](https://pytorch.org/docs/stable/nn.html#crossentropyloss) parameter.

**Comment by sgrvinod** [comment]
@raoanonymous Sorry if it wasn't immediately clear. I've updated the beginning of the [training](https://github.com/sgrvinod/a-PyTorch-Tutorial-to-Image-Captioning#training) and [inference](https://github.com/sgrvinod/a-PyTorch-Tutorial-to-Image-Captioning#inference) sections.

**Comment by RobinWenqian** [comment]
Thanks so much for your advice and really sorry for my late reply. I will try the method that find the similar captions from the dataset by calculating BLEU. If you try the methods that gives discriminating words more weights and get some interesting findings, please inform me. We can then use e-mail to discuss. Mine is Vincent_WQ@outlook.com.

Thanks again for your enthusiasm!

**Comment by sgrvinod** [comment]
Closing for now, reopen if needed, thanks.
